Homebrew 4.0.0 (#929)
Homebrew/brew.sh: _posts/2023-02-07-homebrew-4.0.0.md

---
title: 4.0.0
author: MikeMcQuaid
redirect_from: /blog/4.0.0/
---

> Today I'd like to announce Homebrew 4.0.0. The most significant change since 3.6.0 is faster updating through JSON downloads.

Reviewer: ```suggestion
Today, I'd like to announce Homebrew 4.0.0. The most significant change since 3.6.0 enables significantly faster Homebrew-maintained tap updates by migrating from git-cloned taps to JSON downloads.
```

> Today, I'd like to announce Homebrew 4.0.0. The most significant change since 3.6.0 enables significantly faster Homebrew-maintained tap updates by migrating from git-cloned taps to JSON downloads.

Reviewer: ```suggestion
Today, I'd like to announce Homebrew 4.0.0. The most significant change since 3.6.0 enables significantly faster Homebrew-maintained tap updates by migrating from Git-cloned taps to JSON downloads.
```

Major changes and deprecations since 3.6.0:

- Using JSON files downloaded from [formulae.brew.sh](https://formulae.brew.sh) rather than local homebrew-core and homebrew-cask taps.
  - **Please note: this is the largest change we have made to our update process since we split Homebrew/brew and Homebrew/homebrew-core repositories. Please bear with us, there may be a few bumps**.
  - Unless you are developing formulae or casks, you can `brew untap homebrew/core` and `brew untap homebrew/cask` to save some space.

>   - [`brew update` will now be run automatically less often (every 24 hours rather than every 5 minutes)](https://github.com/Homebrew/brew/pull/14376) and these auto-updates will be much faster as they no longer need to perform the slow `git fetch` of the huge homebrew-core and homebrew-cask taps' Git repositories.

Reviewer: I'd love it if you could provide some concrete metric, at least something like "order of magnitude faster" or some factor like that, even if's entirely synthetic.

Author: Hesistant to do this without actually measuring and can't really get a concrete value for this given the variability.

Reviewer 2: Assuming that network speed for a particular individual is the same when accessing Homebrew repositories, we can provide "back-of-the-envelope" calculations. Here is an example:

Previously, in order to use Homebrew suite on a macOS system, one had to download (and keep updating):

- **Homebrew/brew**
- **Homebrew/core**
- **Homebrew/cask** (on macOS)

After the change, users will have to download (and keep updating) **Homebrew/brew** only and then get tiny jsons.

Here is the breakdown of things:

**Homebrew/brew**: Initial (first-time) full clone: 40s, `.git/` directory: 72 MiB, Commits: 34,000
**Homebrew/core**: Initial (first-time) full clone: 7m 20s, `.git/` directory: 579 MiB, Commits: 346,000
**Homebrew/cask**: Initial (first-time) full clone: 4m, `.git/` directory: 352 MiB, Commits: 170,000

From the above we see that average "commit" sizes are:
- 2.1 KiB in Homebrew/brew
- 1.7 KiB in Homebrew/core
- 2.2 KiB in Homebrew/cask

Repository age:
- 5014 days for Homebrew/brew
- 5000 days for Homebrew/core (🎉, btw)
- 3722 days for Homebrew/cask

So, on average Homebrew repositories produced:
- 7 commits/day by Homebrew/brew
- 70 commits/day by Homebrew/core
- 46 commits/day by Homebrew/cask

Translating that to download size, Homebrew repositories required per day:
- 7 commits/day &times; 2.1 KiB/commit = 14.7 KiB/day
- 70 commits/day &times; 1.7 KiB/commit = 119 KiB/day
- 46 commits/day &times; 2.2 KiB/commit = 101 KiB/day

In summary:
- before the change users had to spend 235 KiB/day.
- after the change users will have to spend 14.7 KiB/day.

This will make things 15 times faster (235/15).

>   - [`HOMEBREW_INSTALL_FROM_API` is now the default behaviour so the variable has been removed and is a no-op](https://github.com/Homebrew/brew/pull/14412).

Reviewer 3: ```suggestion
  - [This is now the default behaviour so the `HOMEBREW_INSTALL_FROM_API` variable has been removed and is a no-op](https://github.com/Homebrew/brew/pull/14412).
```

>   - If you wish to opt-out of this behaviour change, you can [`export HOMEBREW_NO_INSTALL_FROM_API=1`](https://github.com/Homebrew/brew/pull/14375)/

Reviewer: ```suggestion
  - If you wish to opt-out of this behaviour change, you can [`export HOMEBREW_NO_INSTALL_FROM_API=1`](https://github.com/Homebrew/brew/pull/14375)
```

Spurious `/`?

  - This behaviour is automatically disabled when using [certain commands](https://github.com/Homebrew/brew/pull/14524) or in configurations that [mostly build from source](https://github.com/Homebrew/brew/pull/14480).
  - [`HOMEBREW_API_DOMAIN` can be set to use mirrors for](https://github.com/Homebrew/brew/pull/14494) [formulae.brew.sh](https://formulae.brew.sh).
- Work has begun to migrate from Google Analytics to a self-hosted InfluxDB instance. You can [help us test this beta feature by setting `export HOMEBREW_ANALYTICS_ENABLE_INFLUX=1`](https://github.com/Homebrew/brew/pull/13356).

> - [macOS `.pkg` files are generated by releases](https://github.com/Homebrew/brew/pull/14265). You can help us test this beta feature by [downloading the generated package artifact from the relevant GitHub Actions release events](https://github.com/Homebrew/brew/actions/workflows/build-pkg.yml?query=event%3Arelease). [feature: generate macOS pkg files](https://github.com/Homebrew/brew/pull/14265)

Reviewer: ```suggestion
- [macOS `.pkg` files are generated by releases](https://github.com/Homebrew/brew/pull/14265). You can help us test this beta feature by [downloading the generated package artifact from the relevant GitHub Actions release events](https://github.com/Homebrew/brew/actions/workflows/build-pkg.yml?query=event%3Arelease).
```

Drop redundant link

- [The `homebrew/ubuntu16.04:master` image has been deprecated](https://github.com/Homebrew/brew/pull/13819).
- [Various major release deprecations and disables](https://github.com/Homebrew/brew/pull/14382).

Other changes since 3.6.0 I'd like to highlight are the following:

> - [`brew test` sets `PYTHONDONTWRITEBYTECODE` to improve performance.](https://github.com/Homebrew/brew/pull/14338).

Reviewer 3: ```suggestion
- [`brew test` sets `PYTHONDONTWRITEBYTECODE` to improve performance](https://github.com/Homebrew/brew/pull/14338).
```

- [VSCode devcontainers are available in the Homebrew/brew repository for use with e.g. GitHub Codespaces](https://github.com/Homebrew/brew/pull/14027).
- [`brew install --adopt` allows taking ownership of existing installed casks](https://github.com/Homebrew/brew/pull/14033).
- [`brew install --dry-run` allows viewing what `brew install` will do before it is run](https://github.com/Homebrew/brew/pull/13706).
- [`brew docs` opens](https://github.com/Homebrew/brew/pull/13844) [docs.brew.sh](https://docs.brew.sh).
- [`--cask` is never required on Linux](https://github.com/Homebrew/brew/pull/14206).
- [`service do` blocks allow defining a `run` command per platform](https://github.com/Homebrew/brew/pull/14371).
- [`brew install` uses the local cache while installing dependencies from `pip`](https://github.com/Homebrew/brew/pull/14239).
- [`brew doctor` no longer complains about BitDefender](https://github.com/Homebrew/brew/pull/14300).

> - [`brew install` will also suggest casks rather than just formula when it fails to find the requested package](https://github.com/Homebrew/brew/pull/14240).

Reviewer 3: ```suggestion
- [`brew install` will also suggest casks rather than just formulae when it fails to find the requested package](https://github.com/Homebrew/brew/pull/14240).
```

- [`brew readall` simulates all architecture and OS configurations for better reliability](https://github.com/Homebrew/brew/pull/13990).
- [`git` partial clones with sparse checkouts are supported when downloading from `git`](https://github.com/Homebrew/brew/pull/14035).
- [GCC 12 is used for runtime libraries to match Ubuntu 22.04, where we build our bottles](https://github.com/Homebrew/brew/pull/13882).
- [`brew doctor --quiet` prints no output on success](https://github.com/Homebrew/brew/pull/14515).
- [`brew gist-logs` better detects missing permissions](https://github.com/Homebrew/brew/pull/14523).

> - [`brew update` uses the GitHub API token if available to avoid hitting rate-limits](https://github.com/Homebrew/brew/pull/14349).

Reviewer 3: ```suggestion
- [`brew update` uses the GitHub API token if available to avoid hitting rate limits](https://github.com/Homebrew/brew/pull/14349).
```

- [We have discontinued WSL 1 support and recommend WSL 2 instead](https://github.com/Homebrew/brew/pull/14000).
- [`brew fetch` and `brew install` can automatically determine mirrors for `glibc-bootstrap` and PyPI resources](https://github.com/Homebrew/brew/pull/13875).

Finally:

- Homebrew is now on Mastodon: [@homebrew@fosstodon.org](https://fosstodon.org/@homebrew).

> - [Our governance documents have been updated as part of our AGM](https://github.com/Homebrew/brew/pull/14482).

Reviewer: ```suggestion
- [Our governance documents have been updated as part of our AGM](https://github.com/Homebrew/brew/pull/14482). Read the new editions of [Homebrew Governance](https://docs.brew.sh/Homebrew-Governance) and [New Maintainer Checklist](https://docs.brew.sh/New-Maintainer-Checklist) if you're interested.
```

Author: Nah, I think that's probably overkill. Even including it in these notes is tenuous as most users really don't care 😁

> - [We have switched from HackerOne to GitHub for vulnerability reporting](https://github.com/Homebrew/brew/pull/14132).

Reviewer: ```suggestion
- [We have switched from HackerOne to GitHub for vulnerability reporting](https://github.com/Homebrew/brew/pull/14132). Please report security issues [here](https://github.com/homebrew/brew/security/advisories/new).
```

Save someone a click-through maybe.

Author: ```suggestion
- [We have switched from HackerOne](https://github.com/Homebrew/brew/pull/14132) to [GitHub for vulnerability reporting](https://github.com/homebrew/brew/security/advisories/new).
```
